src/Compilers/CSharp/Portable/Errors/DiagnosticInfoWithSymbols.cs

```diff
@@ -25,13 +27,15 @@ internal DiagnosticInfoWithSymbols(bool isWarningAsError, ErrorCode errorCode, o
         public override bool Equals(object obj)
         {
             return obj is DiagnosticInfoWithSymbols diws && 
-                   this.Symbols.Equals(diws.Symbols) && 
+                   (this.Symbols as IStructuralEquatable).Equals(diws.Symbols, EqualityComparer<Symbol>.Default) &&
```

Comment:
I did a search in Roslyn solution and couldn't find a single reference to IStructuralEquatable interface in product code (there are some references in tests). I am sure there are plenty of places where we perform value comparison of two ImmutableArrays and it is very likely that some of them are part of Equals implementation. I recommend to look for those places and follow the same approach. The same goes for GetHashCode.